src/Microsoft.AspNetCore.Mvc.Razor/ViewComponentTagHelperDescriptorFactory.cs

```diff
@@ -0,0 +1,118 @@
+        /// in an <see href="Assembly"/> represented by the given <paramref name="assemblyName"/>.
+        /// </summary>
+        /// <param name="assemblyName">The name of the assembly containing 
+        /// the <see cref="ViewComponent"/>s to translate.</param>
+        /// <returns>A <see cref="IEnumerable{TagHelperDescriptor}"/>, 
+        /// one for each <see cref="ViewComponent"/>.</returns>
+        public IEnumerable<TagHelperDescriptor> CreateDescriptors(string assemblyName)
+        {
+            if (assemblyName == null)
+            {
+                throw new ArgumentNullException(nameof(assemblyName));
+            }
+
+            var viewComponentDescriptors = _descriptorProvider
+                .GetViewComponents()
+                .Where(viewComponent => string.Equals(assemblyName, viewComponent.TypeInfo.Assembly.GetName().Name,
+                    StringComparison.Ordinal));
+
+            var tagHelperDescriptors = viewComponentDescriptors
+                .Select(viewComponentDescriptor => CreateDescriptor(viewComponentDescriptor));
+
+            return tagHelperDescriptors;
+        }
+
+        private TagHelperDescriptor CreateDescriptor(ViewComponentDescriptor viewComponentDescriptor)
+        {
+            var assemblyName = viewComponentDescriptor.TypeInfo.Assembly.GetName().Name;
+            var tagName = GetTagName(viewComponentDescriptor);
+            var typeName = $"__Generated__{viewComponentDescriptor.ShortName}ViewComponentTagHelper";
+
+            var tagHelperDescriptor = new TagHelperDescriptor
+            {
+                TagName = tagName,
+                TypeName = typeName,
+                AssemblyName = assemblyName
+            };
+
+            SetAttributeDescriptors(viewComponentDescriptor, tagHelperDescriptor);
+
+            tagHelperDescriptor.PropertyBag.Add(
+                ViewComponentTagHelperDescriptorConventions.ViewComponentNameKey, viewComponentDescriptor.ShortName);
+
+            return tagHelperDescriptor;
+        }
+
+        private void SetAttributeDescriptors(ViewComponentDescriptor viewComponentDescriptor,
+            TagHelperDescriptor tagHelperDescriptor)
+        {
+            var methodParameters = viewComponentDescriptor.MethodInfo.GetParameters();
+            var attributeDescriptors = new List<TagHelperAttributeDescriptor>();
+
+            foreach (var parameter in methodParameters)
+            {
+                var lowerKebabName = TagHelperDescriptorFactory.ToHtmlCase(parameter.Name);
+                var descriptor = new TagHelperAttributeDescriptor
+                {
+                    Name = lowerKebabName,
+                    PropertyName = parameter.Name,
+                    TypeName = parameter.ParameterType.FullName
+                };
+
+                descriptor.IsEnum = parameter.ParameterType.IsEnum;
+                descriptor.IsIndexer = (parameter.GetType().GetProperties()
```

Comment:
See [here](https://github.com/aspnet/Razor/blob/cdb1a1a3436597523ed4f29e4cebb9bda9d2dfbe/src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperDescriptorFactory.cs#L579) for the logic used to determine indexability of a parameter.